User Contoller 테스트 및 비즈니스 로직 생성

seoko97 opened the pull request:
* Closes #18
* Closes #19 

## ✨ **구현 기능 명세**
- user controller 관련 test 및 비즈니스 로직을 생성했습니다.
- user controller는 총 1개의 함수가 존재하며 내용은 다음과 같습니다.
  - `getUser`:  요청받은 user id에 따라 user를 반환합니다.

## 🎁 **주목할 점**

### `JwtAuthGuard`
- 현재 `user.controller.ts`에선 dummy `JwtAuthGuard`가 설정되어 있습니다. 본래 기능은 허용되지 않은 유저가 접근하는 것을 막는 용도의 decorator가 아닌 무조건 true를 내뱉는 decorator 입니다.
- 해당 부분은 추후 `@nestjs/passport` 라이브러리를 통해 추가적으로 구현해야 합니다.

```ts
// src/guard/jwt-auth.guard.ts
@Injectable()
export class JwtAuthGuard implements CanActivate {
  canActivate(context: any) {
    return true;
  }
  ...
}

// src/api/user/user.controller.ts
@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}
  ...
  @UseGuards(JwtAuthGuard)
  @Get()
  getUser() {
     ...
  }
}
```
### `User Decorator`
- nestjs에서 제공하는 `createParamDecorator`를 통해 Custom Decorator를 생성했습니다.
- 해당 함수의 기능은 후에 생성할 `JwtStrategy`를 통해 context.request에 등록될 user 객체를 가저와 반환하는 Decorator입니다.

```ts
// src/decrators/user.decorator.ts
export const User = createParamDecorator(
  (data: unknown, ctx: ExecutionContext) => {
    const request = ctx.switchToHttp().getRequest();

    if (!request.user) throw new Error('유저 정보가 존재하지 않습니다.');

    return request.user;
  },
);

// src/api/user/user.controller.ts
@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}
  ...
  @UseGuards(JwtAuthGuard)
  @Get()
  getUser(@User() user: UserEntity) {
    return this.userService.getUser(user.id);
  }
}
```

### 테스트 관련
- 위에서 생성한 `JwtAuthGuard`는 테스트에서 정상적으로 동작하지 않습니다.
- 때문에 해당 class를 mocking하여 테스트를 진행합니다.
- 특정 request 요청 완료시 실제 api 응답과 형식을 맞추기 위해 생성된 TestModule에 `ExceptionFilter`, `Interceptor`, `ValidationPipe`를 적용합니다.


```ts
// test/utils/createMockJwtAuthGuard.ts
export const createMockJwtAuthGuard = (id: number) => ({
  canActivate: (context: ExecutionContext) => {
    const request = context.switchToHttp().getRequest();
    request.user = { id };
    return true;
  },
});

// test/api/user/user.controller.spec.ts
describe('UserController', () => {
  const mockUser = Object.assign({}, USER_STUB);
  const mockJwtAuthGuard = createMockJwtAuthGuard(mockUser.id);

  let app: INestApplication;
  let userService: UserService;

  beforeEach(async () => {
    const moduleRef = await createTestingModule(
      {
        controllers: [UserController],
      },
      undefined,
      (builder) => {
        // mocking JwtAuthGuard
        builder.overrideGuard(JwtAuthGuard).useValue(mockJwtAuthGuard);
      },
    );

    app = moduleRef.createNestApplication();

    app.useGlobalFilters(new HttpExceptionFilter());
    app.useGlobalInterceptors(new SuccessInterceptor());
    app.useGlobalPipes(new ValidationPipe({ transform: true }));

    await app.init();

    userService = await app.resolve<UserService>(UserService);
  });
  ...
});

```

## 🌄 **스크린샷**
<img width="410" alt="image" src="https://user-images.githubusercontent.com/60173534/226871059-03e9f2ea-cc5e-4992-ba7b-1a5f22d029b3.png">
- 정상적으로 테스트가 실행됩니다.

seoko97 committed:
feat: builder callback을 통한 함수 확장 (#24)
- createTestingModule.ts에 builder callback를 props로 전달
- builder callback을 통해 기존의 함수를 확장할 수 있도록 함

seoko97 committed:
feat: user controller 관련 stub 데이터 생성 (#18)
- user controller getUser 함수 에러시 결과값에 대한 stub 데이터 생성

seoko97 committed:
feat: user controller test code 작성 (#18)
- getUser: 유저 정보 조회
  - 200: 유저 정보 반환
  - 400: 유저 정보가 없을 경우 에러 반환

seoko97 committed:
feat: createMockJwtAuthGuard 생성 및 적용 (#18)
- controller에서 사용중인 JwtAuthGuard를 mock으로 대체하기 위한 함수 생성

seoko97 committed:
test: createMockJwtAuthGuard 함수 적용 (#18)

seoko97 committed:
feat: JwtAuthGuard frame 생성 (#19)
- 실제로 동작하는 JwtAuthGuard는 아님
- 해당 동작을 위해 필요한 모듈들을 설치하고, JwtAuthGuard를 생성해야함 (@nest/passport 등)

seoko97 committed:
feat: user decorator 생성 (#19)
- 현재 예상중인 플로우에서 user에 대한 정보는 request에 담겨서 전달되는 것으로 예상
- createParamDecorator를 이용하여 request.user에 접근할 수 있도록 구현

seoko97 committed:
feat: user controller getUser 구현 (#19)
- Dummy JwtAuthGuard 적용, 추후 수정 필요
- user decorator를 통한 user 정보 가져오기

jintak0401 commented on src/decorators/user.decorator.ts:
```diff
@@ -0,0 +1,11 @@
+  (data: unknown, ctx: ExecutionContext) => {
+    const request = ctx.switchToHttp().getRequest();
+
+    if (!request.user) throw new Error('유저 정보가 존재하지 않습니다.');
```
그냥 `Error` 보다 `UnauthorizedException`을 사용하는건 어떨까요?

seoko97 replied:
해당 User Decorator는AuthController에서 자세하게 다룰 예정입니다.ᐟ

pyo-sh replied:
저는 Error 자체가 없어도 될 것 같다는 생각이 드네요!
Authorize 관련 작업은 Guard에서 처리해주기 때문에 작업할 것이 없다면 null 과 같은 값으로 반환하도록 만드는게 맞지 않을까요?

jintak0401 commented on test/api/user/user.controller.spec.ts:
```diff
@@ -1,24 +1,97 @@
+      const userServSpy = jest
+        .spyOn(userService, 'getUser')
+        .mockRejectedValue(
+          new BadRequestException('존재하지 않는 유저입니다.'),
```
[BadRequest](https://developer.mozilla.org/ko/docs/Web/HTTP/Status/400)는 요청의 문법상 오류가 있을 때 발생하는 에러라고 합니다. 이보다는 `UnauthorizedException`을 사용하는게 더 나아보입니다.

seoko97 replied:
`UnauthorizedException`는 권한이 존재하지 않는다는 뜻으로 이해됩니다. userService의 `getUser` 잘못된 parameter값이 입력되어 user가 존재하지 않을때 에러를 반환하기 때문에 해당 에러는 `400 BadRequest`가 맞다고 생각합니다.

seoko97 replied:
방금 MDN에 검색해보니 `400 BadRequest`는 잘못된 요청 (형식에 맞지 않는 요청)시 발생하는 에러입니다.
해당 부분은 정상적인 요청중 DB에 자원이 존재하지 않는 것이기 때문에 `404 NotFound`가 맞는 것 같습니다.

jintak0401 replied:
그렇네요. `Not Found`가 제일 맞는 것 같습니다.

jintak0401 commented on src/api/user/user.controller.ts:
```diff
@@ -16,35 +14,12 @@ export class UserController {
-  })
-  @ApiOperation({ summary: '모든 유저정보 데이터 가져오기' })
+  @ApiResponse({ status: 400, description: '유저 정보가 존재하지 않을경우' })
+  @ApiResponse({ status: 200, type: UserEntity })
```
이렇게 사용할 수도 있지만 `ApiOkResponse`, `ApiBadRequestResponse` 를 사용하면 `status` 옵션을 주지 않아도 됩니다! Swagger와 관련된 내용은 Wiki에 작성한 문서가 있으니 확인 한 번 해주세요!

seoko97 replied:
확인했습니다.

jintak0401 reviewed (approved):
고생하셨습니다

seoko97 committed:
feat: user service 에러시 상태 코드 변경 (#18) (#19) (#25)
- getUser: 해당 자원에 대한 유저 정보가 없을 경우 404 상태 코드 반환
- createUser: 이미 존재하는 유저일 경우 409 상태 코드 반환

seoko97 committed:
feat: Swagger 관련 decorator 추가 (#19) (#25)
- 각 에러 코드마다 앚맞은 데코레이터 추가

seoko97 committed:
feat: Swagger ApiOkResponse 추가 (#19) (#25)

jintak0401 commented on test/api/user/user.service.spec.ts:
```diff
@@ -94,7 +95,8 @@ describe('UserService', () => {
         // then
-        expect(error).toBeInstanceOf(BadRequestException);
+        expect(error.status).toEqual(409);
+        expect(error).toBeInstanceOf(ConflictException);
```
별건 아니지만 `expect(error.status).toEqual(409)` 와 `expect(error).toBeInstanceOf(ConfolctException)`이 동일한 것 같습니다. 둘 중에 하나만 있어도 되지 않을까요?

pyo-sh replied:
저는 조금 다르다고 생각합니다.
```javascript
  @Post()
  create(@Res() res: Response) {
    res.status(HttpStatus.CREATED).send();
  }
```
위의 예시 코드와 같이 Nest 에서 Express와 같이 Response 값의 status 값을 직접 정하고 보낼 수 있기 때문입니다.

`expect(error).toBeInstanceOf(ConfolctException)` 를 테스트 코드에 명시했다는 것은 `ConfolctException` 객체를 사용해 에러 처리를 하라는 명령과 같은 것 같습니다

pyo-sh commented on test/utils/createTestingModule.ts:
```diff
@@ -1,29 +1,36 @@
+    }
+  });
+
+  if (builderCb) {
+    builderCb(testingModuleBuilder);
+  }
```
#24 에 관한 내용이군요 !! 좋습니다

pyo-sh commented on test/api/user/user.controller.spec.ts:
```diff
@@ -1,24 +1,95 @@
+        controllers: [UserController],
+      },
+      undefined,
+      (builder) => {
+        builder.overrideGuard(JwtAuthGuard).useValue(mockJwtAuthGuard);
+      },
```
이런 식으로 계속 컨트롤러 테스팅마다 신경써주어야하는게 귀찮기는 하군요 ㅠ
나중엔 관련 builder Mocking 들을 한번에 처리하는 방식도 고려해봐야 겠어요

pyo-sh reviewed (approved):
코드에 고생의 흔적이 보입니다..
화이팅!!
